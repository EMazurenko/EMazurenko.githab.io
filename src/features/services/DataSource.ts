import { AccountType } from 'src/entities/profile/model/types';

class DataSource {
  findAccountDiscount(accountType: AccountType): number {
    throw new Error('Метод findAccountDiscount не реализован');
  }

  findCategoryDiscount(accountType: AccountType, categoryId: string): number {
    throw new Error('Метод findCategoryDiscount не реализован');
  }
}

export default DataSource;
