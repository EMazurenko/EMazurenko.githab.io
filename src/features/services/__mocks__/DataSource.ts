import { AccountType } from 'src/entities/profile/model/types';

type AccountDiscountType = {
  type: AccountType;
  value: number;
};

class DataSource {
  private static readonly ACCOUNT_DISCOUNTS: AccountDiscountType[] = [
    { type: AccountType.STANDARD, value: 0 },
    { type: AccountType.PREMIUM, value: 30 },
    { type: AccountType.GOLD, value: 70 },
    { type: AccountType.FREE, value: 100 },
  ];

  private static readonly CATEGORY_AND_ACCOUNT_DISCOUNTS = [
    {
      category: { id: '1', name: 'Car' },
      accountDiscounts: [
        { type: AccountType.STANDARD, value: 0 },
        { type: AccountType.PREMIUM, value: 0 },
        { type: AccountType.GOLD, value: 40 },
        { type: AccountType.FREE, value: 0 },
      ],
    },
    {
      category: { id: '2', name: 'Toy' },
      accountDiscounts: [
        { type: AccountType.STANDARD, value: 10 },
        { type: AccountType.PREMIUM, value: 10 },
        { type: AccountType.GOLD, value: 30 },
        { type: AccountType.FREE, value: -50 },
      ],
    },
    {
      category: { id: '3', name: 'Food' },
      accountDiscounts: [
        { type: AccountType.STANDARD, value: 15 },
        { type: AccountType.PREMIUM, value: 10 },
        { type: AccountType.GOLD, value: 20 },
      ],
    },
  ];

  findAccountDiscount(accountType: AccountType) {
    return this.findAccountDiscountInCollection(DataSource.ACCOUNT_DISCOUNTS, accountType);
  }

  findCategoryDiscount(accountType: AccountType, categoryId: string) {
    return this.findAccountDiscountInCollection(
      DataSource.CATEGORY_AND_ACCOUNT_DISCOUNTS.filter((discount) => discount.category.id === categoryId)[0]
        ?.accountDiscounts,
      accountType
    );
  }

  private findAccountDiscountInCollection(accountDiscounts: AccountDiscountType[], accountType: AccountType) {
    return accountDiscounts?.filter((discount) => discount.type === accountType)[0]?.value;
  }
}

export default DataSource;
