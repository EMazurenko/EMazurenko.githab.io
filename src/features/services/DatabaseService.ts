import { AccountType } from 'src/entities/profile/model/types';
import DataSource from 'src/features/services/DataSource';

class DatabaseService {
  private readonly dataSource: DataSource;

  constructor(dataSource: DataSource) {
    this.dataSource = dataSource;
  }

  findAccountDiscount(accountType: AccountType) {
    return Math.max(this.dataSource.findAccountDiscount(accountType) || 0, 0);
  }

  findCategoryDiscount(accountType: AccountType, categoryId: string) {
    return Math.max(this.dataSource.findCategoryDiscount(accountType, categoryId) || 0, 0);
  }
}

export default DatabaseService;
