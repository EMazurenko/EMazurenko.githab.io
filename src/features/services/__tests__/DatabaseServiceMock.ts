import DatabaseService from '../DatabaseService';
import DataSource from '../DataSource';
import { ACCOUNT_TYPES, CATEGORIES } from './consts';

jest.mock('../DataSource');

describe('DatabaseServiceMock', () => {
  let databaseService: DatabaseService;

  beforeAll(() => {
    databaseService = new DatabaseService(new DataSource());
  });

  it('should account discounts equals snapshot', () => {
    const discounts = [];
    for (const acc of ACCOUNT_TYPES) {
      discounts.push({
        type: acc,
        value: databaseService.findAccountDiscount(acc),
      });
    }
    expect(discounts).toMatchSnapshot();
  });

  it('should category and account discounts equals snapshot', () => {
    const discounts = [];
    for (const cat of CATEGORIES) {
      for (const acc of ACCOUNT_TYPES)
        discounts.push({
          catId: cat.id,
          accType: acc,
          value: databaseService.findCategoryDiscount(acc, cat.id),
        });
    }
    expect(discounts).toMatchSnapshot();
  });
});
