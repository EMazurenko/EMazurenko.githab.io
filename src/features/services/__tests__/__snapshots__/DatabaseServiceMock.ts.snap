// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatabaseServiceMock should account discounts equals snapshot 1`] = `
[
  {
    "type": 0,
    "value": 0,
  },
  {
    "type": 1,
    "value": 30,
  },
  {
    "type": 2,
    "value": 70,
  },
  {
    "type": 3,
    "value": 100,
  },
]
`;

exports[`DatabaseServiceMock should category and account discounts equals snapshot 1`] = `
[
  {
    "accType": 0,
    "catId": "1",
    "value": 0,
  },
  {
    "accType": 1,
    "catId": "1",
    "value": 0,
  },
  {
    "accType": 2,
    "catId": "1",
    "value": 40,
  },
  {
    "accType": 3,
    "catId": "1",
    "value": 0,
  },
  {
    "accType": 0,
    "catId": "2",
    "value": 10,
  },
  {
    "accType": 1,
    "catId": "2",
    "value": 10,
  },
  {
    "accType": 2,
    "catId": "2",
    "value": 30,
  },
  {
    "accType": 3,
    "catId": "2",
    "value": 0,
  },
  {
    "accType": 0,
    "catId": "3",
    "value": 15,
  },
  {
    "accType": 1,
    "catId": "3",
    "value": 10,
  },
  {
    "accType": 2,
    "catId": "3",
    "value": 20,
  },
  {
    "accType": 3,
    "catId": "3",
    "value": 0,
  },
]
`;
